### DISCUSSION TITLE
中国不缺好的开源开发者 | 夜天之书

### DISCUSSION THREAD

**Discussion by giscus** [tisonkun/blog#94] [Comments]
# 中国不缺好的开源开发者 | 夜天之书

过去数年，我从参与 Perl 6 和 Apache Flink 等项目出发，逐步进入到开源的世界当中。我在 2019 年成为 Apache Flink Committer，2020 年成为 Apache Curator 的 PMC 成员，2021 年全力投入 TiDB 社群的建设，2022 年成为 Apache 软件基金会（ASF）的正式成员和孵化器导师并接连帮助三个开源项目加入孵化器孵化，202

https://www.tisonkun.org/2023/09/16/oss-develop/

**Comment by ZoomQuiet** [comment]
#是也乎,(￣▽￣)
其实, 标题已经自指了对开源的一个误解;
开源从根儿上就不是一个单纯的技术概念,
而是和自由软件一样, 有其政治目标, 只是偏向经济收益机会的...

所以, 好的开源开发者, 其实本身指向的是:
0: 能设立好的开源项目目标
1: 能构建好的开源社区
2: 最关键的, 能良好运营成功一个开源产品获得市场认可

以上三点都作好的, 才是好的开源开发者, 
并不是一堆人好活好又听话的开发者就是好的开源开发者的...
这个世界不能轻易交给那些聪明的坏人,
所以, 好人必须比坏人还要奸诈油滑...

综上, 中国的确缺少好的开源开发者, 因为涉及到的综合能力, 中国传统教育根本就没打算提供...

**Reply by tisonkun** [reply]
> 的确缺少好的开源开发者, 因为涉及到的综合能力

按照你的定义，是这样的。

> 以上三点都作好的, 才是好的开源开发者

确实跟其他人交流过后，我发现这里有一个概念的误解。

1. 我观察到的现象是，有一个好的项目，在发起人有协同意愿（例如 @Xuanwo 的 OpenDAL）的情况下，中国能跟上的开源开发者是不缺的。
2. 缺与不缺看怎么理解，按照我上面的说法，是不缺的。但是有些人解释为他在某个活动场合遇到的参会人员，可能连 Git 都不会用，于是认为缺。这个是没意思的，因为你随机找几个人，不会写代码的也占多数，但是你不能说中国缺会写代码的人。

“都做好”有些太苛责了，我想如果能讲清楚我们不缺基数和能解决问题的人，把关注点放到下面提到的：

1. 引路人，带领更多有能力的开发者进入到开源的世界，了解开源规则。
2. 社群建设，做出好软件的人，如何组织起开源社群（好的软件不是教出来的，至少不是因为开源就有什么大的不同，我们作为一种自然发生的增长，怎么导向开源和开源协同是重要的）。

这样可能更好理解我想强调的点，且这也是我会在今年 ApacheCon 上讲一个几乎没有第二个人讲的《社群长青：开源社群如何可持续》的主题的动因。

**Reply by ZoomQuiet** [reply]
#是也乎,(￣▽￣)
的确, 同时都要能作好, 这个期望本身过于理想化了;

> ...几乎没有第二个人讲的《社群长青：开源社群如何可持续》

结合之前的断言, "可以跟上"...
其实根据以往俺的私人体验来回想, 
一个好的可持续成长的开源社群, 反而并不是一件难的事儿,
比较困难的可能是如何维持长期运营期间核心成员的心态问题吧:

+ 一个 Why not 的想法, 其实很快就可以具像化为一个具体的软件项目
+ 任何一个解决具体问题的软件项目, 当然有很多能跟得上的开发者, 只是要认同最终目标的话, 首先发起团队得指明最终目标(这个也不难, 遵守社区规约合适的时机讨论升级就好, 比如 ECUG 就从 Erlang Chain User Group 很快升级为 Efactive Clud User Group, 缩写没变, 但是, 目标切换了)
+ 一个可持续自我成长的社区, 核心都是一致的, 关注每个成员的成长和项目发展的匹配

核心矛盾在:
无限增长的项目问题矩阵解决过程中核心成员成长速率和其社会资本增长速度不匹配;

也就是说, 虽然在社区中及时回答/解决/引导/...各种具体问题, 带来的自我学习/成长, 一直很快乐,
但是, 对应投入的时间, 相比相同时间精力投入到其它事业中的回报, 对比起来越来越不具备可比性时;
除非:
+ 核心成员快速越过财务自由的心理边界, 可以自由支持时间分配到社区中, 而不影响日益增长的家庭以及其它必要发出压力;
+ 要不核心成员通过良好的社区成长机制, 从新人中不断涌现出来, 替代已经无法享受社区支撑的老核心成员
+ 要不, 有专业成员加入并及时找到开源社区市场成长模式, 核心成员可以将开源项目作为共同事业继续长久投入...

否则, 一但核心成员对项目失去兴趣, 社区又没能及时发现/培养/推举/涌现/...新的核心成员来;
那么, 开源社群以及对应项目自然也就消亡了...

这其实, 也是为什么那些 C*UG 社区, 总是慢慢沉默的原因所在,
一个领域技术的学习社区, 并没有核心项目的持续迭代的话, 很快, 随着核心成员的成长,
后续学习的热情一淡, 新人又无法立即自行继承那种互助的学习氛围...

**Reply by tisonkun** [reply]
> 维持长期运营期间核心成员的心态问题

这就在“可持续”的问题里嘛 :D 我想讲 burn out 的问题的，暂时没找到时间，不过应该优先级还有一些。

下面你说的我基本认同，我上面讲的两点就是某种意义避免单点的做法。引路人是增加基数，后者例如：

StreamPark Top 2 的贡献者 @Al-assad 就曾经因为你说的一些现实问题离开，但是近期也回归了。这里的问题就是在个人离开的时候社群还在不在...

ASF 对项目毕业的要求，大致是 diversity=3，也是这个原因。我在昨天给 OpenDAL 的意见是：

> 可以考虑一下目前各个模块能 cover 的开发者有哪些，再有就是试着写写博客或者讲解讲解 opendal 的内幕。以严格些的毕业的标准来看，目前 opendal 还是某种程度上在漩涡这里单点。
> 
> 比孵化前要好很多，也有不少新的开发者进来，但是 maintainer 层面尤其是不同模块交叉覆盖避免单点，还有提升空间。

当然这个要求是相对严格的。

> 除非

还有一个除非是维护开源软件本身没那么大压力。这个跟“财务自由”有相似的地方，但是也有不一样的地方：国内 996 太厉害，其实正常上下班，不要那么焦虑压力大，维护个把项目问题不大的。乘上基数整个人群就很吓人了。

**Reply by tisonkun** [reply]
> 后续学习的热情一淡, 新人又无法立即自行继承那种互助的学习氛围...

Kvrocks 我很幸运找到了 Twice、mwish 和 torwig 这些积极参与社群的人。当然第一作者 hulk 也一直在 oversee 项目，在我和 hulk 可能不怎么看项目的时候，项目也能正常发展，Twice 能 promote 一些新人。可能还需要提升的就是新的 PMC member 里要有一些人能慧眼识英才。

**Reply by sunheyi6** [reply]
学习热情这个就像是一个事情长远发展的问题，比如公司总是要吸纳年轻人，中国传统文化也是要有人去传承一样，重视年轻人的培养就可以了，当然他们要贡献的基础肯定是要用其他收入来维持自己的日常生活的
